### ISSUE TITLE
Why negelect the output from decoder GRU in second Tutorial

### ISSUE THREAD

**Issue by JanoschMenke** [bentrevett/pytorch-seq2seq#111]
Hi,
I think the Tutorials are great.
I only have a question regarding the useage of the output produced by the decoder GRU in the second Tutorial. It is completely neglected and replaced by the concat. of context, embedding and hidden state. Why is that done?

**Comment by bentrevett** [comment]
In PyTorch, RNNs return an `output` and a `hidden` - the LSTM also returns a `cell` but let's ignore that for now.

When you input a sequence, the `output` is the hidden state output for every single token, obtained by recurrently feeding the hidden state back into the model. The `hidden` is just the last hidden state obtained, i.e. `hidden` == `output[-1,:,:]`.

When we decode, our decode is not fed a sequence, but a single token. Again, the `output` is all of the hidden states and `hidden` is just the final hidden state. But we only have a single token, so `output` and `hidden` are identical. You can prove this by adding an assert statement before the concatenate, i.e.

```python
assert torch.all(torch.eq(output, hidden))
```

Thus, we **are** using the output, as `output` and `hidden` are equal in this scenario. Things get a bit more tricky when using packed padded sequences, bi-directional RNNs and RNNs with multiple layers but the fact that the last element of `output` is equal to `hidden` is always true in some sense, i.e. with multiple layers, `output` is the hidden state only from the last/top RNN layer whereas `hidden` will be the last hidden state from each layer.

**Comment by JanoschMenke** [comment]
Shoot,

yes I forgot we are in the for loop of the decoder.

Could I also ask why there is no non-linearity used after the final layer in the first tutorial. Shouldnt we use a sigmoid on the predictions.

Maybe I missed it but I could not find it.

**Comment by bentrevett** [comment]
The predictions should be fed through a softmax, yes.

PyTorch likes to combine the final activation function and the loss calculating functions together as it allows for more numerically stable calculations. In all of the tutorials we use `nn.CrossEntropyLoss` which actually does both the softmax and the NLL loss together. See: https://pytorch.org/docs/stable/nn.html#torch.nn.CrossEntropyLoss. This means we just need to feed our raw predictions to the criterion (loss function).

**Comment by JanoschMenke** [comment]
Okay, then I was just confused because the nn.BCE does not apply activation beforehand. But the BCEWithLogitLoss does. Thanks you very much.
